results/pkg/api/server/server.go

```diff
@@ -271,11 +264,175 @@ func SetupTestDB(t *testing.T) (*Server, error) {
+				// }
+			} else if e.GetPipelineRun() != nil {
+				log.Printf("Pipeline resolution: %v", e.GetPipelineRun().Metadata.Uid)
+				out, _, err = prg.Eval(map[string]interface{}{
+					"pipelinerun": e.GetPipelineRun(),
+				})
+			} else {
+				continue
+			}
+			if err != nil && !strings.Contains(err.Error(), "no such attribute") && !strings.Contains(err.Error(), "undeclared reference to") {
+				log.Printf("failed to evaluate the expression: %v", err)
+				return nil, 0, status.Errorf(codes.InvalidArgument, "Error occurred during filter evaluation step, no Results found for the query string due to invalid field, invalid function to evaluate filter or missing double quotes around field value, please try to enter a query with correct type again: %v", err)
+			}
+			if out.Value() == true {
+				resp = append(resp, r)
+				got += 1
+				break
+			}
+		}
+		nextIndex = idx + 1
+		if got == requiredSize {
+			break
+		}
+	}
+	return resp, nextIndex, nil
+}
+
+func (s *Server) getFilteredandPagnitatedResult(req *pb.ListResultsRequest, pageSize int32, offset int64, ast *cel.Ast) ([]*pb.Result, string, error) {
+	var (
+		ratio        float32      = 1              // the retention rate after filtering a batch of results.
+		got          int32        = 0              // the number of valid results we got.
+		end          bool         = false          // if there is no more page.
+		batchResults []*pb.Result                  // a batch of results queried from database.
+		results      []*pb.Result = []*pb.Result{} // valid results to return
+	)
+	for got < pageSize {
+		var (
+			nextIndex       int          // the index of the first unchecked result in the batchResults.
+			err             error        // err placeholder
+			filteredResults []*pb.Result // filtered results from the batchResults, its length is no bigger than pageSize-got
+			batchSize       int32        // batchSize = math.Ceil(pageSize/ratio), has a maximum.
+		)
+		if math.Ceil(float64(pageSize)/float64(ratio)) > float64(listResultsDefaultPageSize) {
+			batchSize = listResultsMaximumPageSize
+		} else {
+			batchSize = int32(math.Ceil(float64(pageSize) / float64(ratio)))
+		}
+		batchResults, err = s.getBatchResultsFromDB(batchSize, offset)
+		if err != nil {
+			return nil, "", err
+		}
+		if int32(len(batchResults)) < batchSize {
+			end = true
+		}
+		if req.GetFilter() == "" {
+			filteredResults = batchResults
+			nextIndex = len(batchResults)
+			err = nil
+		} else {
+			filteredResults, nextIndex, err = s.getFilteredResult(batchResults, ast, pageSize-got)
+		}
+		if err != nil {
+			return nil, "", err
+		}
+		results = append(results, filteredResults...)
+		filteredNum := int32(len(filteredResults))
+		if filteredNum != 0 {
+			ratio = float32(filteredNum) / float32(batchSize)
```

Comment:
The ratio heuristic is a bit difficult to reason about as, so at minimum we should add documentation around how this works. IIUC, the idea here is to fetch less items as we approach the number of items we want to return - which is a great idea!

I'm slightly concerned that as we approach the limit, we are trading making more smaller network calls in favor of sending less data. We'll probably want to set a minimum batch size and/or using a simpler selection process to account for this. e.g. would max(remaining results, min batch size) be a sufficient?

Reply:
The key idea is
- the more results remained after filtering, the smaller the next batchSize is.
- no matter how many remained results are, the batchSize only depends on the PageSize and the `ratio`

The following two equations could explain well.
- batchSize = min(PageSize/ratio, max page size)
- ratio = batchFilteredSize/batchSize

Therefore, there wouldn't be the small network calls problem, because the ratio is always less than 1.

The reason why max(remaining results, min batch size) could not work well is, if the ratio is very small, then we may need to query too many times to comprise the final results, and it's hard to determine the value of min batch size.

Hope I expressed myself clearly :)

If needed, I guess we could optimize this function in the future on two aspects:
1. the ratio can be more intelligent(like average over all batch queries).
2. give a maximum of the batchSize(which should be the maximum of the PageSize, the current max size is the maximum of int32)